apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+        end
+    after
+        rocksdb:release_batch(Batch)
+    end.
+
+do_delete(CF, Batch, Static, KeyFamily, MsgKey) ->
+    TS = match_ds_key(Static, MsgKey),
+    lists:foreach(
+        fun({WildcardIdx, Hash}) ->
+            ok = rocksdb:batch_delete(Batch, CF, mk_key(Static, WildcardIdx, Hash, TS))
+        end,
+        KeyFamily
+    ).
+
+%%%%%%%% Iteration %%%%%%%%%%
+
+init_iterators(S, #it{static_index = Static, compressed_tf = CompressedTF}) ->
+    do_init_iterators(S, Static, words(CompressedTF), 1).
+
+do_init_iterators(S, Static, ['+' | TopicFilter], WildcardLevel) ->
+    %% Ignore wildcard levels in the topic filter:
+    do_init_iterators(S, Static, TopicFilter, WildcardLevel + 1);
+do_init_iterators(S, Static, [Constraint | TopicFilter], WildcardLevel) ->
+    %% Create iterator for the index stream:
+    #s{hash_bytes = HashBytes, db = DB, data_cf = DataCF} = S,
+    Hash = hash(HashBytes, Constraint),
+    {ok, ItHandle} = rocksdb:iterator(DB, DataCF, get_key_range(Static, WildcardLevel, Hash)),
+    It = #l{
+        n = WildcardLevel,
+        handle = ItHandle,
+        hash = Hash
+    },
+    [It | do_init_iterators(S, Static, TopicFilter, WildcardLevel + 1)];
+do_init_iterators(S, Static, [], _WildcardLevel) ->
+    %% Create an iterator for the data stream:
+    #s{db = DB, data_cf = DataCF} = S,
+    Hash = <<>>,
+    {ok, ItHandle} = rocksdb:iterator(DB, DataCF, get_key_range(Static, 0, Hash)),
+    [
+        #l{
+            n = 0,
+            handle = ItHandle,
+            hash = Hash
+        }
+    ].
+
+next_loop(
+    Shard,
+    S = #s{trie = Trie},
+    It = #it{static_index = StaticIdx, ts = TS, compressed_tf = CompressedTF},
+    Iterators,
+    BatchSize,
+    TMax
+) ->
+    TopicStructure =
+        case emqx_ds_lts:reverse_lookup(Trie, StaticIdx) of
+            {ok, Rev} ->
+                Rev;
+            undefined ->
+                throw(#{
+                    msg => "LTS trie missing key",
+                    key => StaticIdx
+                })
+        end,
+    Ctx = #ctx{
+        shard = Shard,
+        s = S,
+        iters = Iterators,
+        topic_structure = TopicStructure,
+        filter = words(CompressedTF),
+        tmax = TMax
+    },
+    next_loop(Ctx, It, BatchSize, {seek, TS}, []).
+
+next_loop(_Ctx, It, 0, Op, Acc) ->
+    finalize_loop(It, Op, Acc);
+next_loop(Ctx, It0, BatchSize, Op, Acc) ->
+    %% ?tp(notice, skipstream_loop, #{
+    %%     ts => It0#it.ts, tf => It0#it.compressed_tf, bs => BatchSize, tmax => TMax, op => Op
+    %% }),
+    #ctx{s = S, tmax = TMax, iters = Iterators} = Ctx,
+    #it{static_index = StaticIdx, compressed_tf = CompressedTF} = It0,
+    case next_step(S, StaticIdx, CompressedTF, Iterators, undefined, Op) of
+        none ->
+            %% ?tp(notice, skipstream_loop_result, #{r => none}),
+            inc_counter(?DS_SKIPSTREAM_LTS_EOS),
+            finalize_loop(It0, Op, Acc);
+        {seek, TS} when TS > TMax ->
```

Comment:
Q: I wonder why it's still important to have this stopping criterion here?

Reply:
Future replication layers can use it to communicate high watermark of fully replicated data down the line.